src/unstructured_client/utils/_human_split_pdf.py

```diff
@@ -0,0 +1,137 @@
+import copy
+import io
+import logging
+import os
+import functools
+from typing import Optional, Tuple, Callable
+from concurrent.futures import ThreadPoolExecutor
+
+from pypdf import PdfReader, PdfWriter
+
+from unstructured_client import utils
+from unstructured_client.models import shared, operations
+
+logger = logging.getLogger('unstructured-client')
+
+
+def handle_split_pdf_page(func: Callable) -> Callable:
+    """
+    A decorator for splitting PDF by pages and sending each of them separately to backend.
+    """
+
+    @functools.wraps(func)
+    def wrapper(*args, **kwargs) -> operations.PartitionResponse:
+        if len(args) > 0:
+            request = args[1]
+        else:
+            raise ValueError("Expected a request argument for the partition function.")
+        split_pdf_page = request.split_pdf_page
+
+        if not split_pdf_page:
+            return func(*args, **kwargs)
+
+        pages = get_pdf_pages(request.files.content)
+
+        try:
+            call_threads = int(os.getenv("UNSTRUCTURED_CLIENT_SPLIT_CALL_THREADS", "5"))
+        except ValueError:
+            call_threads = 5
+            logger.error("UNSTRUCTURED_CLIENT_SPLIT_CALL_THREADS has invalid value.")
+        logger.info("Splitting PDF by page on client. Using %d threads when calling API.", call_threads)
+        logger.info("Set UNSTRUCTURED_CLIENT_SPLIT_CALL_THREADS env var if you want to change that.")
+
+        results = []
+        with ThreadPoolExecutor(max_workers=call_threads) as executor:
```

Comment:
can we try with process pool please? multi-threading in python involving models that potentially has their own multi-threading (ocr, layout detection) is dangerous. Additionally this also can trip GIL and cause thread lock.

Reply:
ok, overreacted a bit; this is "just" call api. but the point about potential thread lock stands

Reply:
Hmm, you mean deadlock could happen here? Why? :) I'm not using locks anywhere, and the ThreadPoolExecutor is safe "inside". Maybe you meant potential race condition that can happen in the call_api function which gets sent few shared params:
- request - but it's deep copied to local var, so it's safe for modification.
- retries - those are not copied, I don't modify them and thought partition function also doesn't do it (why would it?) but better to err on the safe side (will also deep copy)
- self - this is sth that I missed! Fortunately looking inside partition, I don't see there any writes, so it seems safe for now, but again, I could deep copy it and it will be safe.

As for processes, those seems unnecesserily heavy for the task here, which is just sending http request - those kinds of IO things are best kept in threads.

Reply:
Done (the deep-copying of all params).